daemon/src/identify/listener.rs

```diff
@@ -0,0 +1,70 @@
+use crate::identify::protocol;
+use crate::Environment;
+use async_trait::async_trait;
+use libp2p_core::Multiaddr;
+use libp2p_core::PublicKey;
+use std::collections::HashSet;
+use tokio_extras::spawn_fallible;
+use xtra::Context;
+use xtra_libp2p::NewInboundSubstream;
+use xtra_productivity::xtra_productivity;
+
+pub struct Actor {
+    daemon_version: String,
+    environment: Environment,
+    identity: PublicKey,
+    listen_addrs: HashSet<Multiaddr>,
+    protocols: HashSet<String>,
+}
+
+impl Actor {
+    pub fn new(
+        daemon_version: String,
+        environment: Environment,
+        identity: PublicKey,
+        listen_addrs: HashSet<Multiaddr>,
+        protocols: HashSet<String>,
+    ) -> Self {
+        Self {
+            daemon_version,
+            environment,
+            identity,
+            listen_addrs,
+            protocols,
+        }
+    }
+}
+
+#[xtra_productivity(message_impl = false)]
+impl Actor {
+    async fn handle(&mut self, message: NewInboundSubstream, ctx: &mut Context<Self>) {
+        let NewInboundSubstream { stream, peer } = message;
+
+        // TODO: Set observed address according to the address we observed when establishing the
+        //  connection
```

Comment:
🔧 So we should fetch it from the `Endpoint` here?

Reply:
We would need the complete multiaddress (i.e. IP + port + p2p, ...) to satisfy this.
I dug into that for a while, but could not figure out how to actually retrieve it from the network stack. It is currently not exposed in our implementation. We would have to find the point in building up the connection where it is available, I was unable to find it. Given that we are not using this information at the moment (and are not planning to in the near future) I would ignore this for now.